pkg/services/pluginsintegration/clientmiddleware/team_headers_middleware.go

```diff
@@ -0,0 +1,175 @@
+	err := m.setHeaders(ctx, req.PluginContext, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.next.CheckHealth(ctx, req)
+}
+
+func (m *TeamHTTPHeadersMiddleware) CollectMetrics(ctx context.Context, req *backend.CollectMetricsRequest) (*backend.CollectMetricsResult, error) {
+	return m.next.CollectMetrics(ctx, req)
+}
+
+func (m *TeamHTTPHeadersMiddleware) SubscribeStream(ctx context.Context, req *backend.SubscribeStreamRequest) (*backend.SubscribeStreamResponse, error) {
+	return m.next.SubscribeStream(ctx, req)
+}
+
+func (m *TeamHTTPHeadersMiddleware) PublishStream(ctx context.Context, req *backend.PublishStreamRequest) (*backend.PublishStreamResponse, error) {
+	return m.next.PublishStream(ctx, req)
+}
+
+func (m *TeamHTTPHeadersMiddleware) RunStream(ctx context.Context, req *backend.RunStreamRequest, sender *backend.StreamSender) error {
+	return m.next.RunStream(ctx, req, sender)
+}
+
+func (m *TeamHTTPHeadersMiddleware) setHeaders(ctx context.Context, pCtx backend.PluginContext, req interface{}) error {
+	reqCtx := contexthandler.FromContext(ctx)
+	// if request not for a datasource or no HTTP request context skip middleware
+	if req == nil || pCtx.DataSourceInstanceSettings == nil || reqCtx == nil || reqCtx.Req == nil {
+		return nil
+	}
+
+	settings := pCtx.DataSourceInstanceSettings
+	jsonDataBytes, err := simplejson.NewJson(settings.JSONData)
+	if err != nil {
+		return err
+	}
+
+	ds := &datasources.DataSource{
+		ID:       settings.ID,
+		OrgID:    pCtx.OrgID,
+		JsonData: jsonDataBytes,
+		Updated:  settings.Updated,
+	}
+
+	signedInUser, err := appcontext.User(ctx)
+	if err != nil {
+		return nil // no user
+	}
+
+	teamHTTPHeaders, err := getTeamHTTPHeaders(ds, signedInUser.GetTeams())
+	if err != nil {
+		return err
+	}
+
+	switch t := req.(type) {
+	case *backend.QueryDataRequest:
+		for key, value := range teamHTTPHeaders {
+			t.SetHTTPHeader(key, value)
+		}
+	case *backend.CheckHealthRequest:
+		for key, value := range teamHTTPHeaders {
+			t.SetHTTPHeader(key, value)
+		}
+	case *backend.CallResourceRequest:
+		for key, value := range teamHTTPHeaders {
+			t.SetHTTPHeader(key, value)
+		}
+	}
+
+	return nil
+}
+
+func getTeamHTTPHeaders(ds *datasources.DataSource, teams []int64) (map[string]string, error) {
+	teamHTTPHeaders := make(map[string]string)
+	teamHTTPHeadersJSON := datasources.TeamHTTPHeadersJSONData{}
+	if ds.JsonData != nil {
+		jsonData, err := ds.JsonData.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(jsonData, &teamHTTPHeadersJSON)
+		if err != nil {
+			return nil, err
+		}
+
+		for teamID, headers := range teamHTTPHeadersJSON.TeamHTTPHeaders {
+			id, err := strconv.ParseInt(teamID, 10, 64)
+			if err != nil {
+				// FIXME: logging here
+				continue
+			}
+			if !contains(teams, id) {
+				continue
+			}
+
+			for _, header := range headers {
+				// TODO: handle multiple header values
+				// add tests for these cases
```

Comment:
I suspect this will be a bit of a challenge for `QueryDataRequest` and `CheckHealthRequest` given that they store headers as a `map[string]string` as opposed to `map[string][]string`.

It seems like the quickest solution would be to fold the headers into a single one with comma-separated values.

Reply:
Yes, this is exactly the way I want to go, since it's RFC-defined way to pass multiple header values